gc/base/ParallelDispatcher.cpp

```diff
@@ -630,3 +644,140 @@ MM_ParallelDispatcher::reinitAfterFork(MM_EnvironmentBase *env, uintptr_t newThr
+		omrthread_monitor_enter(_workerThreadMutex);
+
+		_inShutdown = true;
+
+		/* Clip the thread pool past newThreadCount */
+		for (uintptr_t index = newThreadCount; index < _threadCountMaximum; index++) {
+			_statusTable[index] = worker_status_dying;
+		}
+
+		omrthread_monitor_notify_all(_workerThreadMutex);
+
+		omrthread_monitor_exit(_workerThreadMutex);
+
+		uintptr_t expectedThreadShutdownThread = newThreadCount - 1;
+
+		omrthread_monitor_enter(_dispatcherMonitor);
+		while (expectedThreadShutdownThread != _threadShutdownCount) {
+			omrthread_monitor_wait(_dispatcherMonitor);
+		}
+		omrthread_monitor_exit(_dispatcherMonitor);
+
+		/* Cleanup the dispatcher tables */
+		for(uintptr_t index = newThreadCount; index < _threadCountMaximum; index++) {
+			Assert_MM_true(worker_status_dying == _statusTable[index]);
+			_statusTable[index] = worker_status_inactive;
+			_threadTable[index] = NULL;
+		}
+
+		Assert_MM_true(_threadShutdownCount == expectedThreadShutdownThread);
+
+		_extensions->gcThreadCount = _threadCount = _threadCountMaximum = newThreadCount;
+		_activeThreadCount = adjustThreadCount(_threadCount);
+
+		_inShutdown = false;
+
+		TRIGGER_J9HOOK_MM_PRIVATE_GC_THREAD_POOL_ADJUSTED(_extensions->privateHookInterface, env->getOmrVMThread(), J9HOOK_MM_PRIVATE_GC_THREAD_POOL_ADJUSTED, _extensions->gcThreadCount);
+		Trc_MM_ParallelDispatcher_criu_contractThreadPool_Success(preShutdownThreadCount, newThreadCount);
+	}
+
+	Trc_MM_ParallelDispatcher_criu_contractThreadPool_Exit(_extensions->gcThreadCount);
+}
+
+/**
+ * Expand/fill the thread pool by starting up newThreadCount amount of threads. This API is capped by the initial
+ * thread pool size, i.e we can't expand past _threadCountAtGCInit (currently don't support expanding dispatcher tables).
+ *
+ * This API assumes  CRIU is the only consumer (not tested for general use).
+ * The following conditions are required while expanding the thread pool:
+ *     1) NOT holding exclusive VM access
+ *     2) Dispatcher is idle (no task can be dispatched)
+ *     3) Can't be in/enter shutdown
+ */
+bool
+MM_ParallelDispatcher::expandThreadPool(MM_EnvironmentBase *env, uintptr_t newThreadCount)
+{
+	Trc_MM_ParallelDispatcher_criu_expandThreadPool_Entry();
+
+	Assert_MM_false(_workerThreadsReservedForGC);
+	Assert_MM_false(_inShutdown);
+
+	Assert_MM_true(_threadShutdownCount == (_threadCountMaximum - 1));
+	Assert_MM_true(_threadCountMaximum == _extensions->gcThreadCount);
+
+	bool result = true;
+
+	uintptr_t supportedGCThreadCount = _extensions->configuration->supportedGCThreadCount(env);
+	uintptr_t preRestoreThreadCount = _extensions->gcThreadCount;
+
+	/* Attempt to use H/W supported Thread Count if
+	 * 1) User didn't specify a restore thread count (restoreGCthreadCount == UDATA_MAX), or
+	 * 2) User specified a restore thread count, but it's greater than our hard limit (_threadCountAtGCInit) */
+	if (newThreadCount > _threadCountAtGCInit) {
+		newThreadCount = supportedGCThreadCount;
```

Comment:
this line alone does not make sense (although compound effect with the next line is overall correct)
here we should simply cap to _threadCountAtGCInit
and then on the next line we should again cap to supportedGCThreadCount

but then if we get rid of restoreGCThreadCount, this method should not even pass newThread count and this should be:
set new thread count to _threadCountAtGCInit
cap to supportedGCThreadCount